### PR TITLE
Tweaks the plugin directory search

### PR DESCRIPTION
**Pull request by arcanis**
Basically with Yarn v2 we'll make it possible to load files directly from within the zip archive. For now this will be implemented by patching the fs module so that Node is able to see the files within the archive. The problem is that in order to work well with rimraf and similar crawlers, the archive itself mustn't be stat'd as a directory. This is problematic in Prettier because you actually check it by default.

With this change the node_modules isDirectory check will succeed, and will shortcut the one to check whether its parent directory is a directory.

### CODE REVIEW

**Comment by kachkaev** [comment]
I'm wondering if this new behaviour could be tested somehow 🤔 Otherwise, it won't take long for someone to break it 😅

File: src/common/load-plugins.js [hunk 1]
```diff
@@ -52,17 +52,20 @@ function loadPlugins(plugins, pluginSearchDirs) {
...
-
       const nodeModulesDir = path.resolve(
         resolvedPluginSearchDir,
         "node_modules"
       );
 
+      // In some fringe cases (ex: files "mounted" as virtual directories), the
+      // isDirectory(resolvedPluginSearchDir) check might be false even though
+      // the node_modules actually exists.
+      if (!isDirectory(nodeModulesDir) && !isDirectory(resolvedPluginSearchDir)) {
```
Comment (kachkaev) [inline src/common/load-plugins.js]: From [CI](https://dev.azure.com/prettier/prettier/_build/results?buildId=512):

```
✘  prettier/prettier  Replace `!isDirectory(nodeModulesDir)·&&·!isDirectory(resolvedPluginSearchDir)` with `⏎········!isDirectory(nodeModulesDir)·&&⏎········!isDirectory(resolvedPluginSearchDir)⏎······`  
   src/common/load-plugins.js:63:11
         if (!isDirectory(nodeModulesDir) && !isDirectory(resolvedPluginSearchDir)) {
              ^
 
✘ 1 problem (1 error, 0 warnings)
```

**Comment by arcanis** [comment]
Maybe by adding a mock that would call the real `stat`? Like here:

https://github.com/prettier/prettier/blob/master/tests_integration/runPrettier.js#L54-L56

**Comment by arcanis** [comment]
Added a test 😃

**Comment by kachkaev** [comment]
Linting is [failing again in CI](https://dev.azure.com/prettier/prettier/_build/results?buildId=514):

```
cross-env EFF_NO_LINK_RULES=true eslint . --format node_modules/eslint-friendly-formatter

  ✘  prettier/prettier  Delete `,`                         
  tests_integration/__tests__/plugin-virtual-directory.js:14:14
      write: [],
                ^

  ✘  no-else-return     Unnecessary 'else' after 'return'  
  tests_integration/runPrettier.js:63:12
      } else {
              ^

✘ 2 problems (2 errors, 0 warnings)
```

The first error should be fixed by running this:
```
./node_modules/.bin/prettier --write tests_integration/**/*.*
```

In the second case, the answer seems to be:

```js
    if (path.basename(filename) === `virtualDirectory`) {
      return origStatSync(path.join(__dirname, __filename));
    }
    return origStatSync(filename);
```

**Comment by arcanis** [comment]
Woops, forgot to commit the fix, here it is!

**Comment by arcanis** [comment]
Ping? 🙂

**Comment by j-f1** [comment]
Thanks for contributing!

**Comment by arcanis** [comment]
Any chance this could be released? It would fix a problem that block us from using prettier in the Yarn repo :D

**Comment by j-f1** [comment]
I’m not sure, but in the meantime you can install `prettier/prettier` to pull the latest code from GitHub.

**Pull request merged**